### DISCUSSION TITLE
Migrating from 3.8.x to 3.9.x - Issues with Tax calculation

### DISCUSSION THREAD

**Discussion by satvikt** [saleor/saleor#11564] [General]
Release 3.9.0 includes #9784 , one of the highlights of which is #9999.
I am having a couple of issues with "Flat taxes calculation using tax classes", none of which were present with Vatlayer (until release 3.8.x)

1. [0004_migrate_vatlayer.py](https://github.com/saleor/saleor/blob/1c161a9b2e8d1cebcdbfacad03822433e0704513/saleor/tax/migrations/0004_migrate_vatlayer.py#L42) assumes 

```
Migrate Vatlayer tax codes which are stored in metadata of products and product
types. For each found code get or create a TaxClass instance and assign to the tax
class.
```
this requires presence of `vatlayer.code` as metadata in product and product_type as a pre-requisite. Is this something we need to add manually, before running 0004_migrate_vatlayer.py?

(Since product_types do not have `vatlayer.code` in metadata by default)

As of now, after I run `python manage.py migrate`, I get the following in my saleor local database:

![image](https://user-images.githubusercontent.com/21122115/209481546-5bd74f48-8a29-44a6-af5e-4672884ab3c8.png)

2. [This line](https://github.com/saleor/saleor/blob/240cf66162749c3cd81bb0904381b3ddd5281e6e/saleor/tax/calculations/checkout.py#L35)  seems to return `None` as it searches for those entries in tax_taxclasscountryrate table, where tax_class is None. 

However, as per this [line](https://github.com/saleor/saleor/blob/240cf66162749c3cd81bb0904381b3ddd5281e6e/saleor/tax/migrations/0005_migrate_vatlayer.py#L76) in migration 0005_migrate_vatlayer.py, a TaxClassCountryRate can't exist or be created without a tax_class.

please assist.

**Comment by karolkielecki** [comment]
Hi @satvikt, thanks for reporting. My colleague who was working on those tax changes is currently on holiday. We'll look into it next week and keep you posted.

**Comment by maarcingebala** [comment]
@satvikt The migrations you mentioned are only taken into account if Vatlayer was enabled before migrating to 3.9, otherwise you shouldn't be bothered with them. If there are no Vatlayer's tax codes used, we don't create any tax classes in Saleor. If you want to turn on flat rates calculations, you should create a tax class, add some tax rates and assign this tax class to your products or product types.
